cmd/gateway/zerostor/meta/filesystem_meta.go

```
@@ -0,0 +1,1059 @@
+		if prefixed == path {
+			return nil
+		}
+
+		if err != nil {
+			return errors.Wrapf(err, "failed to walk '%s'", prefixed)
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		if maxKeys == 0 {
+			return errMaxKeyReached
+		}
+
+		name := strings.TrimLeft(strings.TrimPrefix(path, root), "/")
+		if strings.Compare(name, after) <= 0 {
+			//scanning until reach the "after"
+			return nil
+		}
+
+		maxKeys--
+		if !delimited && !strings.HasPrefix(name, prefix) {
+			//match on full prefix and the full object name doesn't have this file/dir
+			return nil
+		}
+
+		last = name
+		if delimited && info.IsDir() {
+			result.Prefixes = append(result.Prefixes, filepath.Clean(name)+"/")
+			return filepath.SkipDir
+		}
+
+		if info.IsDir() {
+			//flat listing (no delmeter)
+			result.Objects = append(result.Objects, minio.ObjectInfo{
+				Bucket: bucket,
+				Name:   filepath.Clean(name) + "/",
+				ETag:   name,
+				IsDir:  true,
+			})
+
+			return nil
+		}
+
+		//we found a file
+		// if file, get metadata of this file
+
+		md, err := m.decodeObjMeta(m.objectFile(bucket, name))
+		if err != nil {
+			return err
+		}
+
+		result.Objects = append(result.Objects, createObjectInfo(bucket, name, &md))
+		return nil
+	})
+
+	return last, err
+}
+
+func (m *filesystemMeta) decodeObjMeta(file string) (ObjectMeta, error) {
+	// open file
+	f, err := os.Open(file)
+	if os.IsNotExist(err) {
+		return ObjectMeta{}, minio.ObjectNotFound{}
+	} else if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			"subsystem": "disk",
+			// "key":       string(key),
+		}).Error("failed to get metadata")
+		return ObjectMeta{}, err
+	}
+
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return ObjectMeta{}, err
+	}
+	if !stat.IsDir() {
+		objMeta := new(ObjectMeta)
+		objMeta.Metadata = *new(metatypes.Metadata)
+		return *objMeta, m.decodeData(f, objMeta)
+	}
+	epoch := stat.ModTime().UnixNano()
+
+	return ObjectMeta{
```

I really think opening a directory should return an error, there is no concept of directories in S3, only prefixes. Hence opening a directory, is similar to opening an object that does not exist.

---

As discussed on telegram, this is used for listing and  is related to how we implement the filesystem meta and the fact that we create directories to represent prefix.